### DISCUSSION TITLE
Scaling multiple objects and automatically resnap sliderends

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#28887] [Q&A]
Let's say I have the following pattern:
![osu_2024-07-17_04-43-45](https://github.com/user-attachments/assets/f8c9b919-60e1-46a6-9866-068fea1df9be)
Now I want to scale the whole pattern to 0.9x. In stable it can be done in this way:
1) add a green line to change sv to 1.8x (2.0*0.9). Now sliderends become unsnapped;
2) select all four objects, ctrl+shift+s (at this moment all sliderends are automatically resnapped);
3) type 0.9 and enter.

In lazer, I tried this:
1) select all four objects, change sv to 1.8x;
2) transform - scale, type 0.9.

However, this leads to unsnapped sliderends:
![osu_2024-07-17_04-47-36](https://github.com/user-attachments/assets/1c4c788c-b3ea-4195-8e57-02f5933cb8f2)

idk how to scale this pattern in lazer editor while keeping sliderends snapped (except, uh, manually adjust slider lengths one by one). Maybe it's useful to add an easy way to resnap sliderends (in stable I can do this by selecting objects, ctrl+shift+s and directly press enter)?

**Comment by peppy** [comment]
To fix this, I would think we need to re-snap distance when slider velocity is changed.

Unfortunately this isn't so simple because the distance snap provider is cached at the composer level, and the popover can't see this. Not immediately sure how we can make this work.

```diff
diff --git a/osu.Game/Rulesets/Objects/SliderPathExtensions.cs b/osu.Game/Rulesets/Objects/SliderPathExtensions.cs
index c03d3646da..cfdeff1999 100644
--- a/osu.Game/Rulesets/Objects/SliderPathExtensions.cs
+++ b/osu.Game/Rulesets/Objects/SliderPathExtensions.cs
@@ -14,10 +14,10 @@ public static class SliderPathExtensions
         /// <summary>
         /// Snaps the provided <paramref name="hitObject"/>'s duration using the <paramref name="snapProvider"/>.
         /// </summary>
-        public static void SnapTo<THitObject>(this THitObject hitObject, IDistanceSnapProvider? snapProvider)
-            where THitObject : HitObject, IHasPath
+        public static void SnapTo(this HitObject hitObject, IDistanceSnapProvider? snapProvider)
         {
-            hitObject.Path.ExpectedDistance.Value = snapProvider?.FindSnappedDistance(hitObject, (float)hitObject.Path.CalculatedDistance) ?? hitObject.Path.CalculatedDistance;
+            if (hitObject is IHasPath pathHitObject)
+                pathHitObject.Path.ExpectedDistance.Value = snapProvider?.FindSnappedDistance(hitObject, (float)pathHitObject.Path.CalculatedDistance) ?? pathHitObject.Path.CalculatedDistance;
         }
 
         /// <summary>
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/DifficultyPointPiece.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/DifficultyPointPiece.cs
index 44235e5d0b..073c547093 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/DifficultyPointPiece.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/DifficultyPointPiece.cs
@@ -16,6 +16,7 @@
 using osu.Game.Graphics;
 using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.UserInterfaceV2;
+using osu.Game.Rulesets.Edit;
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Types;
 using osu.Game.Screens.Edit.Timing;
@@ -69,6 +70,9 @@ public DifficultyEditPopover(HitObject hitObject)
                 this.hitObject = hitObject;
             }
 
+            [Resolved(CanBeNull = true)]
+            private IDistanceSnapProvider distanceSnapProvider { get; set; }
+
             [BackgroundDependencyLoader]
             private void load()
             {
@@ -129,6 +133,8 @@ private void load()
                     foreach (var h in relevantObjects)
                     {
                         ((IHasSliderVelocity)h).SliderVelocityMultiplier = val.NewValue.Value;
+                        h.SnapTo(distanceSnapProvider);
+
                         beatmap.Update(h);
                     }
 

```

**Reply by cihe13375** [reply]
I don't know about the codebase but how is this related to distance snap? The issue is that the sliderends are not snapped (i.e. not on the tick), so if anything it's the sliderends that should be re-snapped?

**Reply by peppy** [reply]
If you don't know the code best to just watch on. Have trust I know what I'm doing.
